### DISCUSSION TITLE
better modulo operation

### DISCUSSION THREAD

**Discussion by MrEgggga** [godotengine/godot-proposals#8140] [Scripting]
currently in gdscript, the modulo operator behaves strangely with negative operands. here's a table showing the results of some modulo operations in gdscript as well as python (where the operator has a different, and in my opinion better, implementation):

| | gdscript | python |
| -: | -: | -: |
| 7 % 3 | 1 | 1 |
| -7 % 3 | -1 | 2 |
| 7 % -3 | 1 | -2 |
| -7 % -3 | -1 | -1 |

to show why the behavior in python is probably more useful, take the example of normalizing an angle into the range `0..360`. in python, this is easily achievable -- to normalize the angle `a` you just calculate `a % 360`. but in gdscript, you have to do something like this:

```gdscript
func normalize_angle(a):
    var mod = a % 360
    if mod < 0:
        mod += 360
    return mod
```

(i suppose `(a % 360 + 360) % 360` is also an option, but that's still 3 unreadable operations compared to one readable one.)

i can't think of any case where gdscript's behavior would be desired.

the cases where the divisor is negative are much less important than the case where the dividend is negative and the divisor is positive, but i personally think the results in python make a lot more intuitive sense.

of course, this is probably a useless suggestion because
- this is a breaking change to gdscript
- it doesn't improve the usability of the engine much at all

**Comment by theraot** [comment]
In GDScript you can do this: `wrapi(a, 0, 360)` or this `posmod(a, 360)`.

**Reply by MrEgggga** [reply]
i was not aware of those functions. thanks

**Comment by theraot** [comment]
GDScript seems to be in good company here: C++, JavaScript, C#... Addendum: Wikipedia has a table on how the operator is implemented in multiple languages: https://en.wikipedia.org/wiki/Modulo#In_programming_languages

Anyway, I found a proposal on the topic: https://github.com/godotengine/godot-proposals/issues/3736

**Comment by dalexeev** [comment]
Thank you for the idea! Let's close this discussion for consolidation (there is already a proposal, as **theraot** said). Note that this breaks compatibility, so cannot be done in 4.x anyway.
